server/src/internalClusterTest/java/org/opensearch/indices/store/IndicesStoreIntegrationIT.java

```diff
@@ -168,6 +168,97 @@ public void testIndexCleanup() throws Exception {
+        Index index = state.metadata().index("test").getIndex();
+
+        logger.info("--> making sure that shard and its replica are allocated on node_1 and node_2");
+        assertThat(Files.exists(shardDirectory(node_1, index, 0)), equalTo(true));
+        assertThat(Files.exists(indexDirectory(node_1, index)), equalTo(true));
+        assertThat(Files.exists(shardDirectory(node_2, index, 0)), equalTo(true));
+        assertThat(Files.exists(indexDirectory(node_2, index)), equalTo(true));
+
+        logger.info("--> starting node server3");
+        final String node_3 = internalCluster().startNode(nonMasterNode());
+        logger.info("--> starting node server4");
+        final String node_4 = internalCluster().startNode(nonMasterNode());
+
+        ClusterHealthResponse clusterHealth = client().admin().cluster().prepareHealth()
+            .setWaitForNodes("5")
+            .setWaitForNoRelocatingShards(true)
+            .get();
+        assertThat(clusterHealth.isTimedOut(), equalTo(false));
+
+        assertThat(Files.exists(shardDirectory(node_1, index, 0)), equalTo(true));
+        assertThat(Files.exists(indexDirectory(node_1, index)), equalTo(true));
+        assertThat(Files.exists(shardDirectory(node_2, index, 0)), equalTo(true));
+        assertThat(Files.exists(indexDirectory(node_2, index)), equalTo(true));
+        assertThat(Files.exists(shardDirectory(node_3, index, 0)), equalTo(false));
+        assertThat(Files.exists(indexDirectory(node_3, index)), equalTo(false));
+        assertThat(Files.exists(shardDirectory(node_4, index, 0)), equalTo(false));
+        assertThat(Files.exists(indexDirectory(node_4, index)), equalTo(false));
+
+        MockTransportService transportService = (MockTransportService) internalCluster().getInstance(TransportService.class, node_4);
+
+        transportService.addMessageListener(new TransportMessageListener() {
+            @Override
+            public void onRequestReceived(long requestId, String action) {
+                if (action.equals(PeerRecoveryTargetService.Actions.FILE_CHUNK)) {
+                    try{
+                        logger.info("received: {}, making relocation stuck", action);
+                        // Dealying relocation.
+                        sleep(500);
+                    } catch (Exception e) {
+                        throw new AssertionError(e);
+                    }
+                }
+            }
+        });
+
+        logger.info("--> move shard from " + node_2 + " to " + node_4 + ", with slow relocation");
+        internalCluster().client().admin().cluster().prepareReroute().add(new MoveAllocationCommand("test", 0, node_2, node_4)).get();
+
+        logger.info("--> move shard from " + node_1 + " to " +  node_3);
+        internalCluster().client().admin().cluster().prepareReroute().add(new MoveAllocationCommand("test", 0, node_1, node_3)).get();
+
+        // so relocation of (node_1 -> node_3) can complete
+        sleep(50);
```

Comment:
Can we use latch here and while delaying relocation to avoid under/over sleep?

Reply:
Adding latch could be little difficult as will need to figure out the place where we can do count down. Instead of latch, I have added while loop to wait untill relocation is not completed and removed this sleep.

Reply:
It is okay to not use latch here, but we should not busy wait either. Even in the while loop have a small sleep like 1 or 5ms, to not burn CPU cycle, but not have any visible slowness for the user.

But, in the transport service mock for node 4, we should use the latch to pause until we need to the validations, etc.